### ISSUE TITLE
Generic Types wrong JSC_TYPE_MISMATCH warning

### ISSUE THREAD

**Issue by g7s** [google/closure-compiler#3471]
This triggers a warning when it shouldn't

```javascript
/**
* @template A,B
*/
class Pair {
  /**
  * @param {A} a
  * @param {B} b
  */
  constructor(a, b) {
      this.a = a;
      this.b = b;
  }
}

class R {}

class T {}

class S {}

/**
* @type {!Pair<T,S>}
*/
R.pair = new Pair(null, null);

exports = R;
```
The warning produced:
```
JSC_TYPE_MISMATCH: assignment to property pair of R
found   : Pair<null,null>
required: Pair<(T|null),(S|null)> at line 26 character 0
R.pair = new Pair(null, null);
```

Is there a workaround to make it work?

**Comment by EatingW** [comment]
Created internal issue http://b/141389103.

**Comment by EatingW** [comment]
Sorry for the delayed response. 

This is actually WAI. The Pair type is invariant. Pair<null, null> is not a subtype of Pair<(T|null), (S|null)>. 
See https://github.com/google/closure-compiler/wiki/Generic-Types#invariance-of-generic-types
